src/synapse_mcp/oauth/client_registry.py

```diff
@@ -0,0 +1,237 @@
+"""Persistence for dynamically registered OAuth clients."""
+
+from __future__ import annotations
+
+from dataclasses import asdict, dataclass
+import json
+import logging
+import os
+from pathlib import Path
+from threading import Lock
+from typing import Iterable, Optional
+
+try:
+    import redis
+    from redis.exceptions import RedisError
+
+    REDIS_AVAILABLE = True
+except ImportError:  # pragma: no cover - redis optional
+    redis = None  # type: ignore[assignment]
+    RedisError = Exception  # type: ignore[assignment, misc]
+    REDIS_AVAILABLE = False
+
+logger = logging.getLogger("synapse_mcp.oauth")
+
+
+@dataclass
+class ClientRegistration:
+    """Serializable representation of a dynamically registered client."""
+
+    client_id: str
+    client_secret: Optional[str]
+    redirect_uris: list[str]
+    grant_types: list[str]
+
+
+class ClientRegistry:
+    """Store for registered OAuth clients."""
+
+    def load_all(self) -> Iterable[ClientRegistration]:  # pragma: no cover - interface only
+        raise NotImplementedError
+
+    def save(self, registration: ClientRegistration) -> None:  # pragma: no cover - interface only
+        raise NotImplementedError
+
+    def remove(self, client_id: str) -> None:  # pragma: no cover - interface only
+        raise NotImplementedError
+
+
+class FileClientRegistry(ClientRegistry):
+    """File-backed registry for DCR clients."""
+
+    def __init__(self, path: Path) -> None:
+        self._path = path
+        self._lock = Lock()
+        self._path.parent.mkdir(parents=True, exist_ok=True)
+
+    def load_all(self) -> list[ClientRegistration]:
+        with self._lock:
+            if not self._path.exists():
+                return []
+            try:
+                data = json.loads(self._path.read_text())
+            except json.JSONDecodeError as exc:  # pragma: no cover - defensive
+                logger.warning("Failed to parse client registry file %s: %s", self._path, exc)
+                return []
```

Comment:
In cases across the MCP server where the exception is being included in the logger string, we can also do something like:

```suggestion
            except json.JSONDecodeError:  # pragma: no cover - defensive
                logger.warning(
                    "Failed to parse client registry file %s.", self._path, exc_info=True)
                return []
```

This will capture the stack trace for the exception automatically. Giving us a lot more context to debug issues as they do pop up.